### PR TITLE
perf: length-bucket inputs in unmask_token_batch (1.9x at bs=32)

### PR DESCRIPTION
**Pull request by AmitMY**
## Summary

Sort prompts by tokenized length and process in fixed-size sub-batches of 8 inside `unmask_token_batch`, then un-sort results before return. The caller-visible API is unchanged — each forward pass just pads up to the longest prompt *in its chunk* rather than the longest in the whole batch.

## Numbers

On GB10, prompt length distribution min=57, p50=104, p95=226, max=262 tokens (512 WordNet examples, 3 warmup / 20 iters):

| batch | baseline ex/s | bucketed ex/s | speedup |
|------:|--------------:|--------------:|--------:|
|   1   | 112 | 109 | 1.0x |
|   4   | 196 | 191 | 1.0x |
|   8   | 178 | 172 | 1.0x *(fits one chunk)* |
|  16   | 151 | 215 | **1.4x** |
|  32   | 124 | 236 | **1.9x** |

- Batches ≤ chunk_size (8) are unaffected — everything fits in one chunk, so there's nothing to sort.
- Larger batches (the common case for one sentence with many content words) see up to 1.9x.
- Without bucketing, throughput *drops* past bs=4 because padding waste dominates; with bucketing, the curve flattens and scales out.

Upper bound for reference: if the entire pool is pre-sorted by length (bench-side `--bucket` flag), bs=8 hits 438 ex/s. Internal bucketing captures most of that on batches > 8 but can't do anything at smaller sizes.

## Changes

- `wsd/masked_language_model.py`: split `unmask_token_batch` into an outer sort/chunk loop and an inner `_unmask_chunk` that does the actual forward pass. Chunk size `_BUCKET_CHUNK_SIZE = 8` chosen empirically.
- `wsd/test_masked_language_model.py`: new `test_unmask_token_batch_preserves_order_with_varied_lengths` exercises the sort path with interleaved short/long prompts; verifies output order matches input order.
- `wsd/bench_inference.py` (new): reproducible HF transformers benchmark so the tradeoff can be re-measured after future changes. Supports `--bucket` to pre-sort the whole pool (establishes the theoretical upper bound).

## Test plan

- [x] `pytest wsd/test_masked_language_model.py` — all 10 tests pass in the CUDA container
- [x] Re-ran `python -m wsd.bench_inference --batch-sizes 1 4 8 16 32 --warmup 3 --iters 20 --n-examples 512` on GB10 and confirmed the numbers above

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes batching behavior and introduces length-based sorting/chunking, which could subtly affect result ordering or edge-case error handling despite tests validating equivalence.
> 
> **Overview**
> Improves `unmask_token_batch` throughput by *length-bucketing* prompts: it now pre-tokenizes to estimate lengths, sorts inputs, runs multiple smaller padded forward passes (chunk size `_BUCKET_CHUNK_SIZE`), then reorders results back to match the original input order.
> 
> Refactors the actual forward-pass logic into `_unmask_chunk`, adds an assertion to ensure all result slots are filled, and introduces a new test that uses varied-length prompts to verify order preservation and consistency vs sequential `unmask_token`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 15d946e237db90ef0b0eea7ebb4aee62d04b4af2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Release Notes

* **New Features**
  * Added a new benchmarking tool to measure inference performance, latency, and throughput across configurable batch sizes and iterations.

* **Performance**
  * Optimized batch processing with length-bucketing to improve inference efficiency.

* **Tests**
  * Added test coverage for batch processing with varied input lengths to ensure correct ordering and consistency.

### CODE REVIEW

File: wsd/masked_language_model.py [hunk 1]
```diff
@@ -101,30 +114,50 @@ def unmask_token_batch(texts: list[str]) -> list[UnmaskResult]:
...
+    # All slots have been filled by the loop above; the cast is just to
+    # satisfy the type checker.
+    return [r for r in results if r is not None]
+
+
+def _unmask_chunk(texts: list[str], components: ModelComponents) -> list[UnmaskResult]:
+    """Single forward pass for a length-homogeneous chunk."""
     inputs = components.tokenizer(texts, return_tensors="pt", padding=True).to(components.device)
 
-    # Find mask token positions for each example in the batch
```
Comment (AmitMY) [inline wsd/masked_language_model.py:107]: do not remove comments unless they are wrong

File: wsd/bench_inference.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
+"""Minimal inference benchmark for the HuggingFace transformers backend.
```
Comment (AmitMY) [inline wsd/bench_inference.py]: can remove this file
